<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/bdm/Daging%20Salai/BigDataHandling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment 6: Mastering Big Data Handling
**Group Name:** Daging Salai <br>
**Group Member:** <br>
Nur Shahirah Binti Jailani (MEC233005) <br>
Mustafa Ghazi Talab (MCS212012)

## TripAdvisor Hotel Reviews
This dataset consists of 878561 reviews (1.3GB) from 4333 hotels crawled from TripAdvisor.

## Loading the Dataset

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [2]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = "/content/gdrive/MyDrive/Assignment 6"

In [3]:
!kaggle datasets download -d joebeachcapital/hotel-reviews

 98% 705M/717M [00:07<00:00, 94.2MB/s]
100% 717M/717M [00:07<00:00, 98.7MB/s]


In [4]:
!unzip \*.zip  && rm *.zip

Archive:  hotel-reviews.zip
  inflating: offering.txt/offering.txt  
  inflating: offerings.csv           
  inflating: review.txt/review.txt   
  inflating: reviews.csv             


In [5]:
!ls

gdrive	offerings.csv  offering.txt  reviews.csv  review.txt  sample_data


## **Strategies for Big Datasets**
### **Load Less Data**

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv('/content/reviews.csv')
df

ratings  \
0       {'service': 5.0, 'cleanliness': 5.0, 'overall'...   
1       {'service': 5.0, 'cleanliness': 5.0, 'overall'...   
2       {'service': 4.0, 'cleanliness': 5.0, 'overall'...   
3       {'service': 5.0, 'cleanliness': 5.0, 'overall'...   
4       {'service': 4.0, 'cleanliness': 5.0, 'overall'...   
...                                                   ...   
878556                                   {'overall': 4.0}   
878557  {'service': 4.0, 'cleanliness': 5.0, 'overall'...   
878558  {'cleanliness': 5.0, 'overall': 5.0, 'rooms': ...   
878559  {'cleanliness': 5.0, 'overall': 5.0, 'rooms': ...   
878560  {'service': 4.0, 'cleanliness': 4.0, 'overall'...   

                                                  title  \
0             “Truly is "Jewel of the Upper Wets Side"”   
1                             “My home away from home!”   
2                                          “Great Stay”   
3                                  “Modern Convenience”   
4       “Its the best of the Andaz Brand in the US....”   
...                                                 ...   
878556                                 “vil komme igen”   
878557                                      “excellent”   
878558                           “Un hotel eccezionale”   
878559                                   “Gerne wieder”   
878560            “Nuevo... novísimo. Estupendo lugar.”   

                                                     text  \
0       Stayed in a king suite for 11 nights and yes i...   
1       On every visit to NYC, the Hotel Beacon is the...   
2       This is a great property in Midtown. We two di...   
3       The Andaz is a nice hotel in a central locatio...   
4       I have stayed at each of the US Andaz properti...   
...                                                   ...   
878556  Pænt hotel i et livligt miljø. Store værelser,...   
878557  un excellent hotel 4 *, pour un prix raisonnab...   
878558  L'hotel, situato in una zona tranquilla e a du...   
878559  Sehr schön ausgestattetes Hotel in bester Lage...   
878560  El hotel está entre el aeropuerto y Washington...   

                                                   author    date_stayed  \
0       {'username': 'Papa_Panda', 'num_cities': 22, '...  December 2012   
1       {'username': 'Maureen V', 'num_reviews': 2, 'n...  December 2012   
2       {'username': 'vuguru', 'num_cities': 12, 'num_...  December 2012   
3       {'username': 'Hotel-Designer', 'num_cities': 5...    August 2012   
4       {'username': 'JamesE339', 'num_cities': 34, 'n...  December 2012   
...                                                   ...            ...   
878556         {'username': '', 'id': '', 'location': ''}            NaN   
878557  {'username': 'dan016', 'num_reviews': 2, 'num_...      July 2008   
878558         {'username': '', 'id': '', 'location': ''}      July 2008   
878559         {'username': '', 'id': '', 'location': ''}      July 2008   
878560  {'username': 'Basajaun', 'num_cities': 8, 'num...     March 2008   

        offering_id  num_helpful_votes        date         id  via_mobile  
0             93338                  0  2012-12-17  147643103       False  
1             93338                  0  2012-12-17  147639004       False  
2           1762573                  0  2012-12-18  147697954       False  
3           1762573                  0  2012-12-17  147625723       False  
4           1762573                  0  2012-12-17  147612823       False  
...             ...                ...         ...        ...         ...  
878556        84093                  0  2008-08-31   51441576       False  
878557        84093                  0  2008-07-18   18003332       False  
878558        84093                  0  2008-04-18   15564515       False  
878559        84093                  0  2008-04-01   15564508       False  
878560        84093                  0  2008-03-13   14259290       False  

[878561 rows x 10 columns]

In [ ]:
df.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878561 entries, 0 to 878560
Columns: 10 entries, ratings to via_mobile
dtypes: bool(1), int64(3), object(6)
memory usage: 1.3 GB


In [ ]:
df1 = df[["ratings", "title"]]
df1.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878561 entries, 0 to 878560
Columns: 2 entries, ratings to title
dtypes: object(2)
memory usage: 253.7 MB


As a result, load less data which is load only required columns has reduce memory usage from 1.3 GB to 253.7 MB

### Use Chunking

- import the libraries

In [ ]:
import pandas as pd
import gc
import glob
import os

In [ ]:
chunk_size= 500000
num= 1
for chunk in pd.read_csv('reviews.csv',chunksize=chunk_size):
  chunk.to_csv('chunk'+str(num)+'.csv', index=False)
  gc.collect()
  num+=1

In [ ]:
df2 =pd.read_csv('chunk2.csv')
df2.head()

ratings  \
0  {'service': 3.0, 'cleanliness': 4.0, 'overall'...   
1  {'service': 1.0, 'cleanliness': 4.0, 'overall'...   
2  {'service': 5.0, 'cleanliness': 5.0, 'overall'...   
3  {'service': 5.0, 'cleanliness': 5.0, 'overall'...   
4  {'service': 5.0, 'cleanliness': 5.0, 'overall'...   

                                               title  \
0    “Luxo prices for less-than-luxo accommodations”   
1  “Front desk terrible, clean and good location ...   
2                 “Big rooms at this boutique hotel”   
3                  “Terrific Hotel – Great Location”   
4            “Had everything we needed... and more!”   

                                                text  \
0  The hotel is in a sort of dodgy area; the plac...   
1  First of all, this review is mostly about the ...   
2  Prompt check in and very friendly staff at the...   
3  We recently returned from San Diego in March. ...   
4  What a lovely hotel. Staff could not do enough...   

                                              author    date_stayed  \
0  {'username': 'CaTraveler7777', 'num_cities': 4...  November 2010   
1  {'username': 'image1', 'num_cities': 3, 'num_h...  November 2010   
2  {'username': 'Mailmanwill', 'num_cities': 24, ...   October 2010   
3  {'username': 'Yakster14', 'num_reviews': 1, 'i...     March 2010   
4  {'username': 'U20manager', 'num_reviews': 1, '...     March 2010   

   offering_id  num_helpful_votes        date        id  via_mobile  
0      1237001                  0  2010-11-16  87332187       False  
1      1237001                  0  2010-11-12  86873317       False  
2      1237001                  0  2010-11-02  85764803       False  
3      1027494                  1  2010-03-28  59767681       False  
4      1027494                  0  2010-03-10  58235443       False

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378561 entries, 0 to 378560
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   ratings            378561 non-null  object
 1   title              378561 non-null  object
 2   text               378561 non-null  object
 3   author             378561 non-null  object
 4   date_stayed        358456 non-null  object
 5   offering_id        378561 non-null  int64 
 6   num_helpful_votes  378561 non-null  int64 
 7   date               378561 non-null  object
 8   id                 378561 non-null  int64 
 9   via_mobile         378561 non-null  bool  
dtypes: bool(1), int64(3), object(6)
memory usage: 26.4+ MB


In [ ]:
for column in df2.columns:
  if df2[column].dtypes == 'int64':
    df2[column] = df2[column].astype('int16')

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378561 entries, 0 to 378560
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   ratings            378561 non-null  object
 1   title              378561 non-null  object
 2   text               378561 non-null  object
 3   author             378561 non-null  object
 4   date_stayed        358456 non-null  object
 5   offering_id        378561 non-null  int16 
 6   num_helpful_votes  378561 non-null  int16 
 7   date               378561 non-null  object
 8   id                 378561 non-null  int16 
 9   via_mobile         378561 non-null  bool  
dtypes: bool(1), int16(3), object(6)
memory usage: 19.9+ MB


As a result, chunk has reduced the memory usage from 2.64 GB to 1.99 GB by subdividing dataset into smaller part.

### **Optimize Data Types**

In [ ]:
%pylab inline
import numpy as np
import pandas as pd
import random
from sys import getsizeof

Populating the interactive namespace from numpy and matplotlib


In [ ]:
df3 = pd.read_csv('/content/reviews.csv')

In [ ]:
start_size = getsizeof(df3)/(1024.0**3)
print('Dataframe size: %2.2f GB'%start_size)

Dataframe size: 1.29 GB


In [ ]:
df3.dtypes

ratings              object
title                object
text                 object
author               object
date_stayed          object
offering_id           int64
num_helpful_votes     int64
date                 object
id                    int64
via_mobile             bool
dtype: object

**Integers**

In [ ]:
df3.offering_id = df3.offering_id.astype('uint8')
df3.num_helpful_votes = df3.num_helpful_votes.astype('uint8')
df3.id = df3.id.astype('uint8')

In [ ]:
print('Dataframe size: %2.2f GB'%(getsizeof(df3)/(1024.0**3)))

Dataframe size: 1.27 GB


**Object**

In [ ]:
date = ['date']
for col in date:
    df3[col] = pd.to_datetime(df3[col])

In [ ]:
print('Dataframe size: %2.2f GB'%(getsizeof(df3)/(1024.0**3)))

Dataframe size: 1.22 GB


**Categories**

In [ ]:
df3.ratings = df3.ratings.astype('category')
df3.title = df3.title.astype('category')
df3.text = df3.text.astype('category')
df3.author = df3.author.astype('category')
df3.date_stayed = df3.date_stayed.astype('category')

In [ ]:
final_size = getsizeof(df3)/(1024.0**3)
print('Dataframe size: %2.2f GB'%final_size)

Dataframe size: 1.16 GB


### Total reduction: 10%!

In [ ]:
print('total size reduction: %2.1f'%((1-final_size/start_size)*100))

total size reduction: 10.0


In [ ]:
df3.dtypes

ratings                    category
title                      category
text                       category
author                     category
date_stayed                category
offering_id                   uint8
num_helpful_votes             uint8
date                 datetime64[ns]
id                            uint8
via_mobile                     bool
dtype: object

As a result, after optimized the Data Types into uint8,category and datetime64, the data frame size has reduce 10% which is from 1.29 GB to 1.16 GB.

### **Sampling**

In [ ]:
import pandas
import random

In [ ]:
df4 = "reviews.csv"
n = sum(1 for line in open(df4))-1  # Calculate number of rows in file
s = n//10  # sample size of 10%
skip = sorted(random.sample(range(1, n+1), n-s))  # n+1 to compensate for header
df5 = pandas.read_csv(df4, skiprows=skip)

<ipython-input-44-5aee92887509>:2: DeprecationWarning: Calling np.sum(generator) is deprecated, and in the future will give a different result. Use np.sum(np.fromiter(generator)) or the python sum builtin instead.
  n = sum(1 for line in open(df4))-1  # Calculate number of rows in file


In [ ]:
df5.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88503 entries, 0 to 88502
Columns: 10 entries, ratings to via_mobile
dtypes: bool(1), int64(3), object(6)
memory usage: 132.6 MB


As a result, after sampling, memory usage has tremendously reduced which is from 1.3 GB to 132.5 MB if used 10% of sample.

### **Parallelize with Dask**

In [7]:
%pylab inline
import os
import pandas as pd
import dask.dataframe as dd
from dask.distributed import Client

Populating the interactive namespace from numpy and matplotlib


In [8]:
!pip install dask

In [9]:
# Reading a Dataframe with Pandas

%%time

df6=pd.read_csv("reviews.csv")

CPU times: user 19.1 s, sys: 3.53 s, total: 22.7 s
Wall time: 30.1 s


In [10]:
# Reading a Dataframe with Dask

%%time

import dask.dataframe as dd
df7=pd.read_csv("reviews.csv")

CPU times: user 16.5 s, sys: 1.73 s, total: 18.2 s
Wall time: 20.1 s


In [11]:
# Merging Dataframe using Pandas

temp1=df6.head(5000)
temp2=df6.tail(5000)

In [12]:
%%time

pdmerge=pd.merge(temp1,temp2)

CPU times: user 36.6 ms, sys: 900 µs, total: 37.5 ms
Wall time: 55.5 ms


In [13]:
# Merging Dataframe using Dask

temp3=df7.head(5000)
temp4=df7.tail(5000)

In [14]:
%%time

ddmerge = dd.merge(temp3,temp4)

CPU times: user 32.1 ms, sys: 805 µs, total: 32.9 ms
Wall time: 34.1 ms


Below is the result after 2 comparison above:

a) Reading Dataframe: Pandas = 30.1s; Dask = 20.1s

b) Merging dataframe: Pandas = 55.5ms; Dask = 34.1ms

## **Comparative Analysis**

**1. Memory usage**
 - Traditional Methods: Batch Processing. Typically involves processing large datasets in batches. Memory usage can be high as the entire batch needs to be loaded into memory for processing.
 - Advanced Strategies: In-Memory Processing. Utilizes in-memory processing techniques where data is kept in RAM, significantly reducing the need to read and write to disk. This can lead to lower memory usage and faster processing.

**2. Computation Time**
  - Traditional Methods: Slower computation time, especially for large datasets.
  - Advanced Strategies: Generally faster due to parallelism, in-memory processing, and real-time capabilities.

## **Advantages of Advances Strategies**
**1. Load Less Data:**
 - Lower Memory Requirements: Minimizes memory usage, particularly beneficial for systems with limited resources.
 - Faster Query Response: Improves query response times by focusing on loading only essential data.

**2. Chunking:**
 - Efficient Storage: Enhances storage efficiency by breaking data into manageable chunks, enabling efficient retrieval.
 - Parallel Processing: Facilitates parallel processing and distributed computing, improving scalability and performance.

**3. Optimize Data Types:**
 - Faster Computations: Optimizes data storage and retrieval, speeding up computations.
 - Memory Efficiency: Reduces memory footprint, allowing for the efficient use of available memory resources.

**4. Sampling:**
 - Reduced Processing Time: Speeds up analytics by working with a representative subset of the data.
 - Resource Efficiency: Reduces the amount of data to be processed, saving computational resources and improving efficiency.

**5. Parallelized with Dask :**
 - Scalability: Enables scalable parallel computing across clusters, accommodating growing data volumes.
 - Resource Optimization: Efficiently utilizes computational resources, leading to improved performance.

## **Conclusion**

In conclusion, the advanced strategies for handling big data, including sampling, chunking, loading less data, parallelized processing with Dask, and optimizing data types, collectively contribute to overcoming the challenges associated with large and complex datasets.

- Enhanced performance. Optimizing data types contribute to improved performance. These strategies reduce memory requirements, speed up data retrieval, and enhance query performance.
- Minimized data transfer and storage costs. Loading less data and optimized data types contribute to reduced data transfer and storage costs. By only working with necessary data subsets and optimizing storage formats, organizations can achieve significant cost savings.
- Scalability and parallel processing. The strategies support scalability by enabling parallel processing. Chunking and parallelized processing with tools like Dask facilitate the distribution of computations across multiple nodes or cores, enhancing speed and efficiency.